src/goto-instrument/synthesizer/synthesizer_utils.cpp

```diff
@@ -0,0 +1,84 @@
+/*******************************************************************\
+
+Module: Utility functions for loop invariant synthesizer.
+
+Author: Qinheping Hu
+
+\*******************************************************************/
+
+#include "synthesizer_utils.h"
+
+goto_programt::targett
+get_loop_end(const unsigned int target_loop_number, goto_functiont &function)
+{
+  natural_loops_mutablet natural_loops(function.body);
+
+  // iterate over all natural loops to find the loop with `target_loop_number`
+  for(const auto &loop_p : natural_loops.loop_map)
+  {
+    const goto_programt::targett loop_head = loop_p.first;
+    goto_programt::targett loop_end = loop_p.first;
+    const loopt &loop = loop_p.second;
+    for(const auto &t : loop)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_end->location_number)
+        loop_end = t;
+    }
+
+    // return if `loop_end` belongs to the target loop
+    if(loop_end->loop_number == target_loop_number)
+      return loop_end;
+  }
+
+  UNREACHABLE;
+}
+
+goto_programt::targett
+get_loop_head(const unsigned int target_loop_number, goto_functiont &function)
+{
+  natural_loops_mutablet natural_loops(function.body);
+
+  // iterate over all natural loops to find the loop with `target_loop_number`
+  for(const auto &loop_p : natural_loops.loop_map)
+  {
+    const goto_programt::targett loop_head = loop_p.first;
+    goto_programt::targett loop_end = loop_p.first;
+    const loopt &loop = loop_p.second;
+    for(const auto &t : loop)
+    {
+      if(
+        t->is_goto() && t->get_target() == loop_head &&
+        t->location_number > loop_end->location_number)
+        loop_end = t;
+    }
+
+    // return if `loop_end` belongs to the target loop
+    if(loop_end->loop_number == target_loop_number)
+      return loop_head;
+  }
+  UNREACHABLE;
+}
```

Comment:
Perhaps you want a single `get_loop` function to factor out the common code of `get_loop_head` and `get_loop_end`?

Reply:
I added a new function ```get_loop_head_or_end``` for the common code.